### ISSUE TITLE
Raspberry PI bookworm - Degraded.

### ISSUE THREAD

**Issue by mikev1963** [elastic/endpoint-package#518]
After installing the elastic agent on a Raspberry pi 4/5 I get the following errors:

```
"@timestamp":"2024-04-29T16:41:38.70703445Z","agent":{"id":"f21a6f26-802c-48ec-91dc-3be609a9fe00","type":"endpoint"},"ecs":{"version":"8.10.0"},"log":{"level":"warning","origin":{"file":{"line":491,"name":"AgentContext.cpp"}}},"message":"AgentContext.cpp:491 Endpoint is setting status to DEGRADED, reason: Policy Application Status","process":{"pid":9828,"thread":{"id":13091}}}
{"@timestamp":"2024-04-29T16:41:58.707882573Z","agent":{"id":"f21a6f26-802c-48ec-91dc-3be609a9fe00","type":"endpoint"},"ecs":{"version":"8.10.0"},"log":{"level":"warning","origin":{"file":{"line":491,"name":"AgentContext.cpp"}}},"message":"AgentContext.cpp:491 Endpoint is setting status to DEGRADED, reason: Policy Application Status","process":{"pid":9828,"thread":{"id":13091}}}
```

This only happens on the Raswpberry PI. I have other Ubuntu servers running x86 that work fine. 

Any help on this agent would be great.  Thanks

**Comment by nicholasberlin** [comment]
@pshef

We support ARM on 5.4+ kernels for recent Ubuntu, SLES, CentOS/RHEL distros.
We don't support Raspberry pis ... though that doesn't mean it won't work.

Based on:
> "Failure enabling process events; current state is disabled."

That means endpoint wasn't able to install event sources, either tracefs based kprobes or ebpf probes.

What kernel version is running? Does the kernel support eBPF, if so does it have btf exported?

You could request a diagnostic package from the `Fleet / Agents` tab and upload it below. 

https://upload.elastic.co/u/7c411cf8-3fb5-4044-ac02-973616fb2ed5 (<--- expires in 7 days)

**Comment by FlorianHeigl** [comment]
@nicholasberlin if you want, make a new upload link please, I got the same on a Raspi CM. It's on an Infineon dev board, so this has:
- same old debian 11.11 (hah)
- kernel version is above 5.4
- CM4 (Lite? I think?)
- 8GB Ram
- 32GB SD

Full kernel string:
`Linux CM4-Lite 5.15.61-v8+ #1579 SMP PREEMPT Fri Aug 26 11:16:44 BST 2022 aarch64 GNU/Linux`

Support bundle is ready
I'll look into eBPF, and try to whatever 'btf' being 'exported' means and report back on that.
I would love to not update soon since the last attempt went so extremely not well. I'll try locate how to rebuild the installed kernel.

Agent details:

```
CPU 
6.76 %
Memory 
385 MB
[View more agent metrics]zzzz
Status
Unhealthy
Last activity
12 seconds ago
Last checkin message
1 or more components/units in a degraded state
Agent ID
4c15a7f9-2a54-483c-8ac2-4f1070ac5ec9
Agent policy
[Base-Linux-Zeek]zzzzz
rev. 7
Agent version
8.19.9
Host name
CM4-Lite
Host ID
zzzzzz
Logging level
info
Privilege mode
Running as root
Agent release
stable
Platform
debian
Monitor logs
Enabled
Monitor metrics
Enabled
```

no cake for the UI formatting there.

i think the most valuable part of this issue is just how unspecific it is reported. i had looked for clear messages anywhere but in the end i could just drop it into google and end up here. in a few weeks maybe I can try it on a larger arm64 system, but basically even if unsupported, basic Pi-like systems are good for validation. It's odd if it doesn't work.

**Comment by nicholasberlin** [comment]
https://upload.elastic.co/u/4af263a2-381b-46dd-a631-607614814be8

@FlorianHeigl

**Comment by FlorianHeigl** [comment]
uploaded
I'll keep notes in case I fix it so I can break it back for any debugging. :-)

**Comment by nicholasberlin** [comment]
## process event capture failed to enable
It failed to load ebpf probes via quark:
```
"2026-04-22T07:43:12.780872746Z 375713 error Watcher.cpp:40 Failed to open quark queue for EBPF event source support check, errno: 95"
```
and failed to install the backup legacy probes (tracefs)
```
"2026-04-22T07:43:14.912089428Z 375713 error PerfWatcher.cpp:176 Failed to write: (p:kprobes/elasticendpoint_PROC_ID_CONNECTOR_probe proc_id_connector which_id=%x1:s32)"
```

## quark
You can investigate [quark](https://github.com/elastic/quark) yourself, it's open source and [well documented](https://elastic.github.io/quark/index.html). The version of endpoint you are running, v8.19.9, uses this commit of quark: https://github.com/elastic/quark/tree/fe065043c96751916d02efecf57f059c750d0fd0

I suggest building and running [`quark-mon -B`](https://elastic.github.io/quark/quark-mon.8.html#B). That should give some pretty clear error output.

## tracefs
My guess is that kernel build did not have [CONFIG_PROC_EVENTS enabled](https://github.com/torvalds/linux/blob/master/drivers/connector/Kconfig).

```
$ grep CONFIG_PROC_EVENTS /boot/config-6.17.0-2*
/boot/config-6.17.0-20-generic:CONFIG_PROC_EVENTS=y
/boot/config-6.17.0-22-generic:CONFIG_PROC_EVENTS=y
```

To get legacy probes working will require a kernel recompile.

**Comment by FlorianHeigl** [comment]
I agree, quark looks understandable. I'll ping back after checking these things.
Gonna try to make it a checklist

# Howto Process Events debugging (WIP)

## 1 Switch kernel

Go to an upstream kernel that isn't missing flags
Probably that's the End

For the rest, as below

## 2 buidling and using the debug container


## 2 eBPF problems

## 2.1 Verify error message exists

if you get an error message like: "Failed to open quark queue for EBPF event source support check"
can be searched in agent logs (add more specific when tested)

###  2.2 Validate for process events eBPF support

- gzip -dc /proc/config.gz | grep CONFIG_PROC_EVENTS` or
- grep CONFIG_PROC_EVENTS /boot/config-`uname -r`* or
- grep CONFIG_PROC_EVENTS /usr/src/linux-headers-`uname -r`
My eval board came with kernel headers and there in the `.config` file I could verify the vendor didn't compile in `CONFIG_PROC_EVENTS`.

### 2.3 Adjust kernel

since this is the current approach (non-legacy) it would be best to fix the issue right there
I can't fill this as of now since (of course) my kernel wasn't packaged,  I only got a dump of headers and need to build it manually later.

in any case, the step here is recompile your kernel (load the old config using i.e. make olddefconfig and then add `CONFIG_PROC_EVENTS=y`. Since Linux isn't a microkernel and this is an internal setting, you need to reboot to apply this. Before rebooting, check that you know how to boot back into your old kernel (i.e. don't make the new one the default, yet)

After reboot, the eBPF process monitoring should enable.


## 3. Legacy fallback

Since this is a two-fold error, let's also see what is wrong with the fallback mode.

### 3.1 see if tracefs is mountable

```
[root@CM4-Lite quark]#  mount -t tracefs nodev /sys/kernel/tracing
mount: /sys/kernel/tracing: nodev already mounted or mount point busy.
```

in that case, check if it's already been mounted (likely by Elastic)

```
[root@CM4-Lite quark]# grep trace /proc/mounts 
tracefs /sys/kernel/tracing tracefs rw,nosuid,nodev,noexec,relatime 0 0
tracefs /sys/kernel/debug/tracing tracefs rw,nosuid,nodev,noexec,relatime 0 0
```

if it's available, we need to dig further (using quark) to find out why it wasn't sufficient as a fallback.



## 4 Buidling quark

you need clang and llvm, bpftool
note, it won't build with version 11 or version 13
(it'll not work with clang11 it doesn't like some parts of Probe.bpf.c

```
# apt update
# apt install clang13 llvm13
# update-alternatives --install /usr/bin/clang clang /usr/bin/clang-13 90
# update-alternatives --install /usr/bin/llvm-strip llvm-strip /usr/bin/llvm-strip-13 90
```

```
$ git clone --recursive https://github.com/elastic/quark
$ cd quark
$ git checkout <tag matching Elastic Agent>
```

```
$ make
```
pray/fix build issues or suppress warnings to the compiler's content.
-> container works



[root@CM4-Lite quark]# make docker-shell
docker run -it -q -v /install/quark:/install/quark -w /install/quark -u 0:0 quark-builder /bin/bash

root@ed717db612b9:/install/quark# /install/quark/quark-mon -b
quark-mon: quark_queue_open: Operation not supported

i feel here some `priviledged` or specific caps for docker are needed?

**Comment by haesbaert** [comment]
Heya!

so you can get a proper error with: `quark-mon -bvvv`, but as someone who has a raspberry pi and had to add BTF support, I would bet my pension it's that:  do you have `/sys/kernel/btf/vmlinux` ? I'm assuming you don't. The fix is sadly to build a custom kernel image with BTF support.

**Comment by FlorianHeigl** [comment]
@haesbaert yeah I had to throw in the towel for the moment, compiling a kernel for my dev board involves like 5 or 6 repos from infineon with various kernels and patches and i don't know. it was just bad. maybe when my health is OK again I can try to get an overview of that mess but for now I just gotta take some notes how far I got it and then forget about it. It was certainly kernel related, that's why I think for anyone with the issue, they are best off to switch to a newer upstream kernel.

Since that board is currently running as 100% sensor role I'll probably just change my agent policy to not include the Linux OS related integrations. It sounds bad even as I type that. But I think I can use the auditd integration as a reasonable replacement (monitor only)

**Comment by haesbaert** [comment]
> [@haesbaert](https://github.com/haesbaert) yeah I had to throw in the towel for the moment, compiling a kernel for my dev board involves like 5 or 6 repos from infineon with various kernels and patches and i don't know. it was just bad. maybe when my health is OK again I can try to get an overview of that mess but for now I just gotta take some notes how far I got it and then forget about it. It was certainly kernel related, that's why I think for anyone with the issue, they are best off to switch to a newer upstream kernel.
> 
> Since that board is currently running as 100% sensor role I'll probably just change my agent policy to not include the Linux OS related integrations. It sounds bad even as I type that. But I think I can use the auditd integration as a reasonable replacement (monitor only)

Heya, I'll see if I find some time to resurrect my raspberry pi and then I can send you the kernel image, I built it months ago.

**Comment by FlorianHeigl** [comment]
@haesbaert thanks, but don't bother - it'd miss the wifi drivers that cause the fuss when wanting to switch kernels. just gotta wait for the right day and then i'll try again :-)
